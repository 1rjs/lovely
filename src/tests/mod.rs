use std::fs;

use insta::glob;

use crate::{checker::Checker, ir::IRGenerator, lexer::Lexer, parser::Parser};

#[test]
fn compiler_tests() {
    glob!("source_files/*.lv", |path| {
        let input = fs::read_to_string(path).unwrap();

        // test the lexer
        let lexer = Lexer::new(&input);
        let tokens = lexer.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);

        // test the parser
        let mut parser = Parser::new(&input);
        let ast = parser.parse().unwrap();
        insta::assert_debug_snapshot!(ast);

        // test the checker
        let mut checker = Checker::new();
        let checked_program = checker.check(&ast).unwrap();
        insta::assert_debug_snapshot!(checked_program);

        // test the IR
        let ir_generator = IRGenerator::new(checker.types);
        let ir = ir_generator.program_ir(&checked_program);
        let ir_string = ir
            .iter()
            .map(|b| b.to_string())
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(ir_string);
    });
}
