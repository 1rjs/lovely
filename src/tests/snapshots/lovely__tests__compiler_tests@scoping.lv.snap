---
source: src/tests/mod.rs
expression: tokens
input_file: src/tests/source_files/scoping.lv
---
[
    Token {
        kind: Identifier,
        span: Span {
            start: 0,
            end: 1,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 2,
            end: 3,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 3,
            end: 4,
        },
    },
    Token {
        kind: LBrace,
        span: Span {
            start: 5,
            end: 6,
        },
    },
    Token {
        kind: Identifier,
        span: Span {
            start: 9,
            end: 10,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 11,
            end: 12,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 12,
            end: 13,
        },
    },
    Token {
        kind: LBrace,
        span: Span {
            start: 14,
            end: 15,
        },
    },
    Token {
        kind: Identifier,
        span: Span {
            start: 20,
            end: 21,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 22,
            end: 23,
        },
    },
    Token {
        kind: Colon,
        span: Span {
            start: 23,
            end: 24,
        },
    },
    Token {
        kind: IntLiteral,
        span: Span {
            start: 25,
            end: 26,
        },
    },
    Token {
        kind: Identifier,
        span: Span {
            start: 31,
            end: 32,
        },
    },
    Token {
        kind: Minus,
        span: Span {
            start: 33,
            end: 34,
        },
    },
    Token {
        kind: IntLiteral,
        span: Span {
            start: 35,
            end: 37,
        },
    },
    Token {
        kind: RBrace,
        span: Span {
            start: 40,
            end: 41,
        },
    },
    Token {
        kind: Identifier,
        span: Span {
            start: 44,
            end: 45,
        },
    },
    Token {
        kind: Plus,
        span: Span {
            start: 46,
            end: 47,
        },
    },
    Token {
        kind: IntLiteral,
        span: Span {
            start: 48,
            end: 50,
        },
    },
    Token {
        kind: RBrace,
        span: Span {
            start: 51,
            end: 52,
        },
    },
    Token {
        kind: Identifier,
        span: Span {
            start: 54,
            end: 55,
        },
    },
    Token {
        kind: Minus,
        span: Span {
            start: 56,
            end: 57,
        },
    },
    Token {
        kind: IntLiteral,
        span: Span {
            start: 58,
            end: 59,
        },
    },
]
